components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/CORSRequestHandler.java

```diff
@@ -16,69 +16,191 @@
+			allowedOrigins = Arrays.asList(ServiceReferenceHolder.getInstance().getAPIManagerConfiguration().
+					getFirstProperty(APIConstants.CORS_CONFIGURATION_ACCESS_CTL_ALLOW_ORIGIN).split(","));
+		}
+		return true;
+	}
+
+	public void destroy() {
+		if (log.isDebugEnabled()) {
+			log.debug("Destroying CORSRequest Handler handler instance");
+		}
+	}
+
+	public boolean handleRequest(MessageContext messageContext) {
+		if (!headerStatus) {
+			headerStatus = initializeHeaders();
+		}
+		String apiContext = (String) messageContext.getProperty(RESTConstants.REST_API_CONTEXT);
+		String apiVersion = (String) messageContext.getProperty(RESTConstants.SYNAPSE_REST_API_VERSION);
+		String httpMethod = (String) ((Axis2MessageContext) messageContext).getAxis2MessageContext().
+				getProperty(Constants.Configuration.HTTP_METHOD);
+		API selectedApi = null;
+		Resource selectedResourceWithVerb = null;
+		Resource selectedResource = null;
+		boolean status = false;
+
+		for (API api : messageContext.getConfiguration().getAPIs()) {
+			if (apiContext.equals(api.getContext()) && apiVersion.equals(api.getVersion())) {
+				selectedApi = api;
+				break;
+			}
+		}
+		String subPath;
+		String path = RESTUtils.getFullRequestPath(messageContext);
+		if (selectedApi.getVersionStrategy().getVersionType().equals(VersionStrategyFactory.TYPE_URL)) {
+			subPath = path.substring(
+					selectedApi.getContext().length() + selectedApi.getVersionStrategy().getVersion().length() + 1);
+		} else {
+			subPath = path.substring(selectedApi.getContext().length());
+		}
+		if ("".equals(subPath)) {
+			subPath = "/";
+		}
+		messageContext.setProperty(RESTConstants.REST_SUB_REQUEST_PATH, subPath);
+
+		if (selectedApi.getResources().length > 0) {
+			for (RESTDispatcher dispatcher : RESTUtils.getDispatchers()) {
+				Resource resource = dispatcher.findResource(messageContext, Arrays.asList(selectedApi.getResources()));
+				if (resource != null) {
+					selectedResource = resource;
+					if (Arrays.asList(resource.getMethods()).contains(httpMethod)) {
+						selectedResourceWithVerb = resource;
+						break;
+					}
+				}
+			}
+		}
+		String resourceString =
+				selectedResourceWithVerb != null ? selectedResourceWithVerb.getDispatcherHelper().getString() : null;
+		String resourceCacheKey = APIUtil
+				.getResourceInfoDTOCacheKey(apiContext, apiVersion, resourceString, httpMethod);
+		messageContext.setProperty(APIConstants.API_ELECTED_RESOURCE, resourceString);
+		messageContext.setProperty(APIConstants.API_RESOURCE_CACHE_KEY, resourceCacheKey);
+		setCORSHeaders(messageContext, selectedResourceWithVerb);
+		if (selectedResource != null && selectedResourceWithVerb != null) {
+				if ("inline".equals(inline)) {
+					messageContext.getSequence("_cors_request_handler").mediate(messageContext);
+				}
+				status =  true;
+			}else if (selectedResource != null && selectedResourceWithVerb == null ){
+			if ("OPTIONS".equalsIgnoreCase(httpMethod)) {
```

Comment:
Remember to use constants and avoid string when applicable. Since this is a http method we can move to constants calss. Or refer it from available constant class same thing can done for _cors_request_handle which is referred from  multiple places.  Next time you can follow that :+1: